### PR TITLE
ENH: Add non-execution dask code

### PR DESCRIPTION
**Pull request by gerrymanoim**
Part 2/4 of https://github.com/ibis-project/ibis/issues/2537

Notes:
- I've imported from the pandas backend when doable
- I haven't added aggcontext since that wasn't working yet 
- I've xfailed a few tests that are awaiting implementation of `execution` coming in part 3. 
- I haven't added udf yet since most of the test code is awaiting `execution`.

### CODE REVIEW

File: ibis/backends/dask/core.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+           return it
+       Else:
+           execute the arguments of the current node
+
+       execute the current node with its executed arguments
+
+Specifically, execute is comprised of a series of steps that happen at
+different times during the loop.
+
+1. ``compute_time_context``
```
Comment (jreback) [inline ibis/backends/dask/core.py:45]: line under this

File: ibis/backends/dask/core.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+contract of window functions: given N rows of input there are N rows of output.
+
+Defining a ``post_execute`` rule for :class:`~ibis.expr.operations.WindowOp`
+allows you to encode such logic. One might want to implement this using
+:class:`~ibis.expr.operations.ScalarParameter`, in which case the ``scope``
+passed to ``post_execute`` would be the bound values passed in at the time the
+``execute`` method was called.
+
+
+Scope
```
Comment (jreback) [inline ibis/backends/dask/core.py:95]: nitpick same length as the title

File: ibis/backends/dask/core.py [hunk 3]
```diff
@@ -0,0 +1,544 @@
...
+# Register is_computable_input for each scalar type (int, float, date, etc).
+# We use consume here to avoid leaking the iteration variable into the module.
+ibis.util.consume(
+    is_computable_input.register(t)(is_computable_input_arg)
+    for t in scalar_types
+)
+
+
+def execute_with_scope(
+    expr,
```
Comment (jreback) [inline ibis/backends/dask/core.py:176]: type expr

File: ibis/backends/dask/core.py [hunk 4]
```diff
@@ -0,0 +1,544 @@
...
+    scope: Scope,
+    timecontext: Optional[TimeContext] = None,
```
Comment (jreback) [inline ibis/backends/dask/core.py:178]: make a todo item to combine a number of these execution primitives for pandas/dask

Comment (jreback) [reply ibis/backends/dask/core.py:178]: though you *might* simply be able to import the pandas ones directly

Comment (jreback) [reply ibis/backends/dask/core.py:178]: yeah definitely try to do this here (if it gets bogged down, then just create a todo)

File: ibis/backends/dask/tests/test_client.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    assert 'DaskTable' in repr(table)
+
+
+def test_load_data(client):
+    client.load_data('testing', make_dask_data_frame())
+    assert client.exists_table('testing')
+    assert client.get_schema('testing')
+
+
+@pytest.mark.xfail(reason="Need to fix up create_table")
```
Comment (jreback) [inline ibis/backends/dask/tests/test_client.py:57]: can you add the TODO issue (can be same one for anything you are xfailing / marking TODO)

File: ibis/backends/dask/tests/test_core.py [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+from .. import connect, execute, from_dataframe
+from ..client import DaskClient
+from ..core import is_computable_input
+from ..dispatch import execute_node, post_execute, pre_execute
+
+pytestmark = pytest.mark.dask
+
+
+@pytest.fixture
```
Comment (jreback) [inline ibis/backends/dask/tests/test_core.py:23]: can you move fixtures to conftest.py (I think there is one already)?

File: ibis/backends/dask/tests/test_datatypes.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        (np.float32(5.5), dt.float),
+        (np.float64(5.55), dt.float64),
+        (np.float64(5.55), dt.double),
+        (np.bool_(True), dt.boolean),
+        (np.bool_(False), dt.boolean),
+        (np.arange(5, dtype='int32'), dt.Array(dt.int32)),
+        # pandas types
+        (
+            pd.Timestamp('2015-01-01 12:00:00', tz='US/Eastern'),
+            dt.Timestamp('US/Eastern'),
```
Comment (jreback) [inline ibis/backends/dask/tests/test_datatypes.py:40]: add some TODO / issues for more modern types like Period/Interval/Int; you can test Categorical as well (though this might be covered elsewhere)

File: ibis/backends/dask/tests/test_datatypes.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+def test_dtype_bool():
+    df = dd.from_pandas(
+        pd.DataFrame({'col': [True, False, False]}), npartitions=1
+    )
+    inferred = sch.infer(df)
+    expected = ibis.schema([('col', 'boolean')])
+    assert inferred == expected
+
+
+def test_dtype_int8():
```
Comment (jreback) [inline ibis/backends/dask/tests/test_datatypes.py:106]: paramaterize these (do here, can do pandas backend in separate PR if easier)

File: ibis/backends/dask/tests/test_dispatcher.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+class A3(A2):
+    pass
+
+
+class B1(object):
+    pass
+
+
+class B2(B1):
+    pass
```
Comment (jreback) [inline ibis/backends/dask/tests/test_dispatcher.py:25]: i actually don't think this is useful as the dispatcher is already tested in pandas, remove unless there is specific customization (do this generally)

File: ibis/backends/dask/tests/test_schema.py [hunk 10]
```diff
@@ -0,0 +1,215 @@
+import dask.dataframe as dd
+import numpy as np
```
Comment (jreback) [inline ibis/backends/dask/tests/test_schema.py:2]: this is heavily overlapping with test_datatypes.py on purpose?

File: ibis/backends/dask/trace.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+        f"{datetime.now() - start if start else ''}"
+    )
+
+
+def trace(func):
+    """ Return a function decorator that wraped the decorated function with
+    tracing.
+    """
+    _trace_funcs.add(func.__name__)
+
```
Comment (jreback) [inline ibis/backends/dask/trace.py:111]: is this needed at all?

we can simply move the dispatcher / trace to a common module (maybe worthile doing that first).

you should be able to put in ibis/common (with the assocaited tests).

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. however would really try to re-use as much as possible. the dispatcher / trace look like just imports and can directly use the pandas (or move to ibis/common if needed). the execution machinery might be more tricky to de-couple (so ok on doing this later). but do as much as possible now.

File: ibis/backends/dask/__init__.py [hunk 12]
```diff
@@ -10,3 +20,63 @@
...
         validator=ibis.config.is_bool,
     )
+
+
+def connect(dictionary):
+    """Construct a dask client from a dictionary of DataFrames.
+
+    Parameters
+    ----------
+    dictionary : dict
```
Comment (icexelloss) [inline ibis/backends/dask/__init__.py:30]: Can you please explain is the the `dictionary ` here? Also feels like this could be a bit better named

File: ibis/backends/dask/__init__.py [hunk 13]
```diff
@@ -10,3 +20,63 @@
...
+
+    Returns
+    -------
+    DaskClient
+    """
+    return DaskClient(dictionary)
+
+
+def from_dataframe(df, name='df', client=None):
```
Comment (icexelloss) [inline ibis/backends/dask/__init__.py:39]: I would suggest removing the default value "df" for `name` and make it required. 

If "client" is None, it seems like you created a Client object inside, but how does the user get an handler to that client?

File: ibis/backends/dask/client.py [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+    ibis_dtype_to_pandas,
+    ibis_schema_to_pandas,
+    infer_array,
+    infer_numpy_scalar,
+    infer_pandas_timestamp,
+)
+
+from .core import execute_and_reset
+
+infer_dask_dtype = pd.api.types.infer_dtype
```
Comment (icexelloss) [inline ibis/backends/dask/client.py:42]: Does dask just pandas's dtypes for data types and schema?

File: ibis/backends/dask/client.py [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+
+
+ibis_dtype_to_dask = ibis_dtype_to_pandas
+
+ibis_schema_to_dask = ibis_schema_to_pandas
+
+convert = Dispatcher(
+    'convert',
+    doc="""\
+Convert `column` to the dask dtype corresponding to `out_dtype`, where the
```
Comment (icexelloss) [inline ibis/backends/dask/client.py:96]: "dask dtype" -> "dask Series"?

File: ibis/backends/dask/client.py [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+dtype of `column` is `in_dtype`.
+
+Parameters
+----------
+in_dtype : Union[np.dtype, dask_dtype]
+    The dtype of `column`, used for dispatching
+out_dtype : ibis.expr.datatypes.DataType
```
Comment (icexelloss) [inline ibis/backends/dask/client.py:103]: Do we need this? Looks like out_dtype is always dd.Series

File: ibis/backends/dask/client.py [hunk 17]
```diff
@@ -0,0 +1,342 @@
...
+        pairs.append((column_name, ibis_dtype))
+
+    return sch.schema(pairs)
+
+
+ibis_dtype_to_dask = ibis_dtype_to_pandas
+
+ibis_schema_to_dask = ibis_schema_to_pandas
+
+convert = Dispatcher(
```
Comment (icexelloss) [inline ibis/backends/dask/client.py:93]: Can you perhaps give this a better name? Looks like just this method is used to cast the type of dask Series?

File: ibis/backends/dask/core.py [hunk 18]
```diff
@@ -0,0 +1,544 @@
...
+    # TODO: make expresions hashable so that we can get rid of these .op()
+    # calls everywhere
+    params = {k.op() if hasattr(k, 'op') else k: v for k, v in params.items()}
+    scope = scope.merge_scope(Scope(params, timecontext))
+    return execute_with_scope(
+        expr, scope, timecontext=timecontext, aggcontext=aggcontext, **kwargs,
+    )
+
+
+def execute_and_reset(
```
Comment (icexelloss) [inline ibis/backends/dask/core.py:439]: Does this method actually trigger the dask job? Or does it just return a deferred dask.DataFrame that is executed later?

File: ibis/backends/dask/trace.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+)
+def add_one(v):
+    import time
+    time.sleep(5)
+    return v + 1
+table = con.table("table1")
+table = table.mutate(b=add_one(table['a']))
+table.execute()
+Output:
+DEBUG:ibis.dask.trace: main_execute Selection
```
Comment (icexelloss) [inline ibis/backends/dask/trace.py:41]: I don't think the tracing make sense for dask if this is lazily executed...

File: ibis/backends/dask/dispatch.py [hunk 20]
```diff
@@ -0,0 +1,11 @@
+from ibis.backends.pandas.dispatch import (  # noqa: F401
+    execute,
+    execute_literal,
```
Comment (icexelloss) [inline ibis/backends/dask/dispatch.py:3]: How does this work? These pandas dispatch methods are defined to take and return pandas Series. It looks like you are reusing this to return dd.Series? How does that work?

**Review by icexelloss — COMMENTED** [review]
Took a initial quick look for architecture. Left some comments.

File: ibis/backends/dask/dispatch.py (hunk 20 above)
Comment (icexelloss) [reply ibis/backends/dask/dispatch.py:3]: Can you explain, perhaps, how does this work for Dask backend?

https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/selection.py#L372

I believe you dispatch into this but this method has code like

```
    # create a sequence of columns that we need to drop
    temporary_columns = pd.Index(
        concatv(grouping_keys, ordering_keys)
    ).difference(data.columns)

    # no reason to call drop if we don't need to
    if temporary_columns.empty:
        return result

    # drop every temporary column we created for ordering or grouping
    return result.drop(temporary_columns, axis=1)
```
How does this code work for dask Series as is?

**Comment by icexelloss** [comment]
In general, I am +1 to what is here. Although, I main comments about the design are:

* Instead of copying backends/pandas/core.py, we should to reuse it as is for dask backend (perhaps with slight modification). I believe core.py just describes the general bottom up tree execution and should be pandas/dask agnostic. 
* I have doubts about how much of backends/pandas/dispatch can we actually reuse, those logic are written for pd.Series and pd.DataFrame in general and use the code "as-is" for dask.Series and dask.DataFrame seems problematic. For example, in

https://github.com/ibis-project/ibis/blob/master/ibis/backends/pandas/execution/selection.py#L313

There are code like:
```
        new_pieces = [
            piece.reset_index(
                level=list(range(1, piece.index.nlevels)), drop=True
            )
            if piece.index.nlevels > 1
            else piece
            for piece in data_pieces
        ]
        result = pd.concat(new_pieces, axis=1)
```

and 
```
    # return early if we do not have any temporary grouping or ordering columns
    assert not grouping_keys, 'group by should never show up in Selection'
    if not ordering_keys:
        return result

    # create a sequence of columns that we need to drop
    temporary_columns = pd.Index(
        concatv(grouping_keys, ordering_keys)
    ).difference(data.columns)

    # no reason to call drop if we don't need to
    if temporary_columns.empty:
        return result
```
that are pandas specific. Try to use this logic on dask.Series seems problematic. We should probably have dask specific version of these dispatch rules and perhaps the super simple ones can be shared, like the rule for "+" operator.

**Comment by jreback** [comment]
to @icexelloss point. Yeah ideally we want to share as much of the pandas impl as possible, if we can re-use dispatch rules let's just do it. Certainly in places we might need to have separate rules.

At the same time it is definintely worth having a full test suite in place. So ideally let's push forward to have coverage on most things (we can leave the TODO's about aggregation / windowing for later).

Then refactor to combine things in smaller PRs is ok.

File: ibis/backends/dask/client.py [hunk 21]
```diff
@@ -0,0 +1,342 @@
...
+from pandas.api.types import DatetimeTZDtype
+from pkg_resources import parse_version
+
+import ibis.client as client
+import ibis.common.exceptions as com
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+import ibis.expr.schema as sch
+import ibis.expr.types as ir
+from ibis.backends.pandas.client import (  # noqa: F401
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:23]: I think we need these? They're to eliminate "import without using" errors for parts we're borrowing for the pandas backend.

File: ibis/backends/dask/tests/test_client.py (hunk 5 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/test_client.py:57]: Do you prefer new issues, or can I continue to update https://github.com/ibis-project/ibis/issues/2537 with more detail?

File: ibis/backends/dask/tests/test_dispatcher.py (hunk 9 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/test_dispatcher.py:25]: I duplicated the tests between pandas and dask so that if the dask pieces ever change we'll know if there's a functionality break. I can drop any tests that are testing pandas functionality if that's your preference though.

File: ibis/backends/dask/tests/test_schema.py (hunk 10 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/test_schema.py:2]: Nope - just ported.

File: ibis/backends/dask/client.py (hunk 14 above)
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:42]: Yep

File: ibis/backends/dask/core.py (hunk 18 above)
Comment (gerrymanoim) [reply ibis/backends/dask/core.py:439]: As much as possible the dask backend does not run `compute` and returns a dask expression that the user can choose to execute. Sometimes it is unavoidable, but that's the goal.

File: ibis/backends/dask/trace.py (hunk 19 above)
Comment (gerrymanoim) [reply ibis/backends/dask/trace.py:41]: I think it still helps to trace what ibis is doing?

File: ibis/backends/dask/dispatch.py (hunk 20 above)
Comment (gerrymanoim) [reply ibis/backends/dask/dispatch.py:3]: As requested, I haven't implemented anything inside `execution` for the dask backend yet (that's part 3 of https://github.com/ibis-project/ibis/issues/2537). You can see what the `execution` pieces would look like in https://github.com/ibis-project/ibis/pull/2400. We use the pandas machinery, but files under `execution` will register dask types where appropriate.

**Comment by gerrymanoim** [comment]
Thanks for the comments - I'll clean up the above. I believe we can reuse most of the pandas machinery, but all the code under `execution` (coming in the next PR) will be dask specific. With the exception of various aggregations, most of the logic is very similar, but registered against dask types. 

> the dispatcher / trace look like just imports and can directly use the pandas (or move to ibis/common if needed). the execution machinery might be more tricky to de-couple (so ok on doing this later). but do as much as possible now.

Sorry - I'm a bit unlcear. Taking `TwoLevelDispatcher` as an example, do you prefer that `ibis.backends.dask.dispatcher` import `pandas.dispatcher.TwoLevelDispatcher` and then `ibis.backends.dask.trace` imports `from .dispatcher`? Or should trace import directly from pandas? I think the former is a bit cleaner, but happy to do either.

File: ibis/backends/dask/dispatch.py (hunk 20 above)
Comment (icexelloss) [reply ibis/backends/dask/dispatch.py:3]: Ok cool. I wasn't sure what is working and what is not working yet. 

Sounds like no node execution rule (e.g., rule for Selection) is implemented in this PR yet?

File: ibis/backends/dask/trace.py (hunk 19 above)
Comment (icexelloss) [reply ibis/backends/dask/trace.py:41]: Yeah that's true. The pandas tracing tool was original written to debug performance, I guess you can use it for just getting the stack trace but it's less useful that way.

Either way, I would prefer not to duplicate all the tracing logics, but we can address as follow up.

File: ibis/backends/dask/client.py (hunk 14 above)
Comment (icexelloss) [reply ibis/backends/dask/client.py:42]: kk, good to know. Thanks!

File: ibis/backends/dask/dispatch.py (hunk 20 above)
Comment (gerrymanoim) [reply ibis/backends/dask/dispatch.py:3]: Yep - exactly. I had implemented it in https://github.com/ibis-project/ibis/pull/2400, but 11k lines is too big of a PR to actually review so based on Jeff's suggestion I'm splitting this into 4-6 smaller PRs. Part 1 was the skeleton, Part 2 (this PR) borrows the pandas machinery to build up the dask client, and Part 3 (upcoming) will actually deal with the `execution` components.

Comment (icexelloss) [reply ibis/backends/dask/dispatch.py:3]: Got it. Sounds good to me.

File: ibis/backends/dask/trace.py (hunk 19 above)
Comment (gerrymanoim) [reply ibis/backends/dask/trace.py:41]: Yep - I believe I can import most of it from pandas.

**Comment by icexelloss** [comment]
Side note: I think the entirety of the pyspark backend is around 2K lines of code (mostly this file https://github.com/ibis-project/ibis/blob/master/ibis/backends/pyspark/compiler.py)

My hope is that the complexity of the Dask backend is close to that XD

**Comment by gerrymanoim** [comment]
With the exception of pieces that can be used from pandas, I expect the length/structure of the dask backend to be similar to the pandas backend. I think that should be on the order of 2kish lines (excluding tests).

File: ibis/backends/dask/__init__.py (hunk 12 above)
Comment (jreback) [reply ibis/backends/dask/__init__.py:30]: this is already present in all other backends (agreed its not great)

File: ibis/backends/dask/__init__.py [hunk 22]
```diff
@@ -10,3 +20,63 @@
         'See ibis.dask.trace for details.',
         validator=ibis.config.is_bool,
     )
+
+
+def connect(dictionary):
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/__init__.py:25]: Do you want registered functions typed as well? For example:

```
@sch.schema.register(dd.Series)
def schema_from_series(s: dd.Series):
```

File: ibis/backends/dask/core.py (hunk 4 above)
Comment (gerrymanoim) [reply ibis/backends/dask/core.py:179]: Yeah - you're right - not sure why I didn't import these previously.

File: ibis/backends/dask/__init__.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jreback) [reply ibis/backends/dask/__init__.py:25]: no that's a much big undertaking

we should do it but as a todo (on a single issue list or make a separate issue doesn't matter)

File: ibis/backends/dask/client.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:23]: I think as a user of the library you'd expect pandas/dask to have the same interface, so it makes sense to have these imported from pandas but unused, so users could import them from dask without worrying about which pieces come from pandas and which pieces are specific to dask. Also gives a reasonable starting point for a future PR to consolidate things into common. I can change this if you disagree though.

File: ibis/backends/dask/trace.py (hunk 11 above)
Comment (gerrymanoim) [reply ibis/backends/dask/trace.py:111]: This currently exists separately so that it has its own logger (`ibis.dask.trace`) and `_TRACE_CONFIG`  (`'dask.enable.trace`). 

I agree that this should get consolidated eventually, don't have a strong view on what we do with it now. Right now dask.core is importing all the pandas machinery anyway, so I'm happy to get rid of trace.py/dispatcher.py for now.

File: ibis/backends/dask/__init__.py (hunk 13 above)
Comment (gerrymanoim) [reply ibis/backends/dask/__init__.py:39]: I agree in principle with your suggestions, but I think it makes sense to keep the top level interfaces of dask and pandas backends the same. I can create a new issue to fix this in both backends together?

I think you can grab the client off of the `source` variable on the table objects, but agreed that's not user friendly.

File: ibis/backends/dask/client.py (hunk 16 above)
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:103]: I think the `column` type is always `dd.Series`, while the input/output dtypes vary.

Comment (icexelloss) [reply ibis/backends/dask/client.py:104]: Yeah I meant sincne column is also dd.Series, we should probably remove it from the dispatch signature?

File: ibis/backends/dask/core.py (hunk 18 above)
Comment (icexelloss) [reply ibis/backends/dask/core.py]: kk I understand this now - "execute" does not actually trigger computation, but return a dd.DataFrame/Series instead.

**Comment by icexelloss** [comment]
| With the exception of pieces that can be used from pandas, I expect the length/structure of the dask backend to be similar to the pandas backend. I think that should be on the order of 2kish lines (excluding tests).

Yeah this sounds very reasonable.

File: ibis/backends/dask/client.py (hunk 16 above)
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:104]: Ah - I see what you're saying. I need to double check why we have this interface in the pandas backend, maybe to support the possibilities of other outputs in the future?

File: ibis/backends/dask/core.py [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+stored in value
+
+See ibis.common.scope for details about the implementaion.
+"""
+
+from __future__ import absolute_import
+
+import dask.dataframe as dd
+
+from ibis.backends.pandas.core import (  # noqa: F401
```
Comment (jreback) [inline ibis/backends/dask/core.py:115]: same here, let's try to remove the noqa and only import things we are actually using.

File: ibis/backends/dask/tests/test_client.py [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+import pandas as pd
+import pandas.util.testing as pandas_tm
+import pytest
+from dask.dataframe.utils import tm
+from pytest import param
+
+import ibis
+
+from .. import connect
+from ..client import DaskTable  # noqa: E402
```
Comment (jreback) [inline ibis/backends/dask/tests/test_client.py:12]: really try to remove all noqa

File: ibis/backends/dask/tests/test_client.py (hunk 5 above)
Comment (jreback) [reply ibis/backends/dask/tests/test_client.py:57]: i think have a tracking issue is good here (with the check boxes)

File: ibis/backends/dask/tests/test_dispatcher.py (hunk 9 above)
Comment (jreback) [reply ibis/backends/dask/tests/test_dispatcher.py:25]: ok, make a task to factor out the dispatching to a common routine somewhere

File: ibis/backends/dask/trace.py [hunk 25]
```diff
@@ -0,0 +1,147 @@
...
+import logging
+import traceback
+from datetime import datetime
+
+from ibis.config import get_option, set_option
+from ibis.expr import types as ir
+
+from .dispatcher import TwoLevelDispatcher
+
+"""Module that adds tracing to dask execution.
```
Comment (jreback) [inline ibis/backends/dask/trace.py:11]: same with this, pls create an issue, we want to factor out to a common location (ok for now)

File: ibis/backends/dask/trace.py (hunk 11 above)
Comment (jreback) [reply ibis/backends/dask/trace.py:111]: yeah ok to do later

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good. a few followup tasks to note and some changes would like here. ping on green-ish

File: ibis/backends/dask/trace.py (hunk 11 above)
Comment (gerrymanoim) [reply ibis/backends/dask/trace.py:111]: https://github.com/ibis-project/ibis/issues/2551

File: ibis/backends/dask/trace.py (hunk 25 above)
Comment (gerrymanoim) [reply ibis/backends/dask/trace.py:11]: https://github.com/ibis-project/ibis/issues/2551

File: ibis/backends/dask/client.py [hunk 26]
```diff
@@ -0,0 +1,354 @@
...
+    return column.astype(out_dtype.to_dask(), errors='ignore')
+
+
+DASK_STRING_TYPES = PANDAS_STRING_TYPES
+DASK_DATE_TYPES = PANDAS_DATE_TYPES
+
+
+@convert.register(np.dtype, dt.Timestamp, dd.Series)
+def convert_datetime64_to_timestamp(in_dtype, out_dtype, column):
+    if in_dtype.type == np.datetime64:
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:131]: https://github.com/ibis-project/ibis/issues/2552

File: ibis/backends/dask/client.py [hunk 27]
```diff
@@ -0,0 +1,354 @@
...
+
+@convert.register(np.dtype, dt.String, dd.Series)
+def convert_any_to_string(_, out_dtype, column):
+    result = column.astype(out_dtype.to_dask(), errors='ignore')
+    return result
+
+
+@convert.register(np.dtype, dt.Boolean, dd.Series)
+def convert_boolean_to_series(in_dtype, out_dtype, column):
+    # XXX: this is a workaround until #1595 can be addressed
```
[the remark this answers is not in the document]
Comment (gerrymanoim) [reply ibis/backends/dask/client.py:169]: https://github.com/ibis-project/ibis/issues/1595

Came into pandas as part of https://github.com/ibis-project/ibis/commit/b8dd27644049efaa2ff58d16bbeba54860b4bd31 and was ported directly to dask.

File: ibis/backends/dask/tests/test_client.py (hunk 5 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/test_client.py:57]: https://github.com/ibis-project/ibis/issues/2553

**Comment by gerrymanoim** [comment]
I've dealt with the noqas and got rid of of dispather/dispatch, importing from the pandas backend where appropriate. When I do the execution pieces, I'll move the imports there.

File: ibis/backends/dask/tests/test_datatypes.py (hunk 7 above)
Comment (gerrymanoim) [reply ibis/backends/dask/tests/test_datatypes.py:40]: https://github.com/ibis-project/ibis/issues/2554

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @gerrymanoim very nice.
